# 18 ARTIFICIAL NEURAL NETWORKS 
Case Study: SONAR — Detecting Mines vs. Rocks

### 1️⃣ Business Objective
#### Goal:
To build an intelligent system that can automatically detect whether an underwater sonar signal is reflected from a metallic mine (potentially dangerous) or a harmless rock.
This is vital for:
- Maritime safety: Prevent ships and submarines from colliding with mines.
- Naval defense: Identify and safely remove underwater mines.
- Resource exploration: Distinguish between useful metal structures and natural seabed objects.


### 2️⃣ Problem Statement
In underwater environments, sonar (sound navigation and ranging) is used to detect objects. However, raw sonar signals can be noisy and difficult for humans to interpret consistently.
#### This dataset:
- Contains 208 sonar returns.
    - 111 are from metal cylinders (mines).
    - 97 are from rocks.
- Each sonar return is represented by 60 numeric features, each measuring the energy of the signal in a frequency band.

#### The problem:
👉 To train a Deep learning model that can learn the difference in signal patterns and classify new sonar signals as either Mine (M) or Rock (R) — accurately and reliably.


### Dataset: "sonardataset.csv"
#### Features (Inputs)
- There are 60 numerical variables, each representing the energy in a specific frequency band of the sonar signal.
- In the original dataset, they’re just unnamed columns V1, V2, ..., V60 — you can keep it clear and simple:
#### Target (Output)
- The label is a single categorical variable indicating:
    - "M" for Mine
    - "R" for Rock


### Tasks :
#### 1. Data Exploration and Preprocessing
- Begin by loading and exploring the "Alphabets_data.csv" dataset. Summarize its key features such as the number of samples, features, and classes.
- Execute necessary data preprocessing steps including data normalization, managing missing values.


In [1]:
import pandas as pd
import numpy as np 
import warnings
warnings.filterwarnings("ignore")

df = pd.read_csv("sonardataset.csv")
df.head()

x_1     x_2     x_3     x_4     x_5     x_6     x_7     x_8     x_9  \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
3  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
4  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   

     x_10  ...    x_52    x_53    x_54    x_55    x_56    x_57    x_58  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
3  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
4  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   

     x_59    x_60  Y  
0  0.0090  0.0032  R  
1  0.0052  0.0044  R  
2  0.0095  0.0078  R  
3  0.0040  0.0117  R  
4  0.0107  0.0094  R  

[5 rows x 61 columns]

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x_1     208 non-null    float64
 1   x_2     208 non-null    float64
 2   x_3     208 non-null    float64
 3   x_4     208 non-null    float64
 4   x_5     208 non-null    float64
 5   x_6     208 non-null    float64
 6   x_7     208 non-null    float64
 7   x_8     208 non-null    float64
 8   x_9     208 non-null    float64
 9   x_10    208 non-null    float64
 10  x_11    208 non-null    float64
 11  x_12    208 non-null    float64
 12  x_13    208 non-null    float64
 13  x_14    208 non-null    float64
 14  x_15    208 non-null    float64
 15  x_16    208 non-null    float64
 16  x_17    208 non-null    float64
 17  x_18    208 non-null    float64
 18  x_19    208 non-null    float64
 19  x_20    208 non-null    float64
 20  x_21    208 non-null    float64
 21  x_22    208 non-null    float64
 22  x_

In [3]:
df.duplicated().sum()

0

In [4]:
df.isnull().value_counts()

x_1    x_2    x_3    x_4    x_5    x_6    x_7    x_8    x_9    x_10   x_11   x_12   x_13   x_14   x_15   x_16   x_17   x_18   x_19   x_20   x_21   x_22   x_23   x_24   x_25   x_26   x_27   x_28   x_29   x_30   x_31   x_32   x_33   x_34   x_35   x_36   x_37   x_38   x_39   x_40   x_41   x_42   x_43   x_44   x_45   x_46   x_47   x_48   x_49   x_50   x_51   x_52   x_53   x_54   x_55   x_56   x_57   x_58   x_59   x_60   Y    
False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False  False    208
Name: count, dtype: int64

In [5]:
df['Y'].value_counts()

Y
M    111
R     97
Name: count, dtype: int64

In [6]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df["Y"] = le.fit_transform(df["Y"])

In [7]:
df.head(3)

x_1     x_2     x_3     x_4     x_5     x_6     x_7     x_8     x_9  \
0  0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109   
1  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
2  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   

     x_10  ...    x_52    x_53    x_54    x_55    x_56    x_57    x_58  \
0  0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084   
1  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
2  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   

     x_59    x_60  Y  
0  0.0090  0.0032  1  
1  0.0052  0.0044  1  
2  0.0095  0.0078  1  

[3 rows x 61 columns]

In [8]:
X = df.drop("Y" , axis = 1)
y = df["Y"]

In [9]:
from sklearn.preprocessing import MinMaxScaler
mms = MinMaxScaler()
x_scaled = mms.fit_transform(X)

### 2. Model Implementation
- Construct a basic ANN model using your chosen high-level neural network library. Ensure your model includes at least one hidden layer.
- Divide the dataset into training and test sets.
- Train your model on the training set and then use it to make predictions on the test set.

In [10]:
from  sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x_scaled, y, test_size = 0.2 , random_state = 42 , stratify = y)

In [11]:
from sklearn.neural_network import MLPClassifier
#Build & Train ANN Model
ann = MLPClassifier(
    hidden_layer_sizes=(16, 8),
    activation='relu',
    solver='adam',
    max_iter=500,
    random_state=42,
    verbose=True,
)

ann.fit(X_train, y_train)

Iteration 1, loss = 0.69664787
Iteration 2, loss = 0.69294966
Iteration 3, loss = 0.68961043
Iteration 4, loss = 0.68657108
Iteration 5, loss = 0.68362715
Iteration 6, loss = 0.68085705
Iteration 7, loss = 0.67807355
Iteration 8, loss = 0.67531154
Iteration 9, loss = 0.67253253
Iteration 10, loss = 0.66972858
Iteration 11, loss = 0.66694036
Iteration 12, loss = 0.66413192
Iteration 13, loss = 0.66145557
Iteration 14, loss = 0.65879917
Iteration 15, loss = 0.65608126
Iteration 16, loss = 0.65346887
Iteration 17, loss = 0.65087406
Iteration 18, loss = 0.64821521
Iteration 19, loss = 0.64552550
Iteration 20, loss = 0.64278924
Iteration 21, loss = 0.63997907
Iteration 22, loss = 0.63711687
Iteration 23, loss = 0.63422411
Iteration 24, loss = 0.63125040
Iteration 25, loss = 0.62819899
Iteration 26, loss = 0.62507888
Iteration 27, loss = 0.62197905
Iteration 28, loss = 0.61897367
Iteration 29, loss = 0.61594880
Iteration 30, loss = 0.61299400
Iteration 31, loss = 0.61005942
Iteration 32, los

MLPClassifier(hidden_layer_sizes=(16, 8), max_iter=500, random_state=42,
              verbose=True)

In [12]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, classification_report
# Evaluation
y_pred = ann.predict(X_test)

print("\n--- BASIC ANN PERFORMANCE ---")
print("Accuracy :", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall   :", recall_score(y_test, y_pred))
print("F1 Score :", f1_score(y_test, y_pred))

print("\nClassification Report:")
print(classification_report(y_test, y_pred))


--- BASIC ANN PERFORMANCE ---
Accuracy : 0.9047619047619048
Precision: 1.0
Recall   : 0.8
F1 Score : 0.8888888888888888

Classification Report:
              precision    recall  f1-score   support

           0       0.85      1.00      0.92        22
           1       1.00      0.80      0.89        20

    accuracy                           0.90        42
   macro avg       0.92      0.90      0.90        42
weighted avg       0.92      0.90      0.90        42



### 3. Hyperparameter Tuning
- Modify various hyperparameters, such as the number of hidden layers, neurons per hidden layer, activation functions, and learning rate, to observe their impact on model performance.
- Adopt a structured approach like grid search or random search for hyperparameter tuning, documenting your methodology thoroughly.

In [13]:
param_grid = {
    'hidden_layer_sizes': [(32,), (64,), (100,), (64,32)],
    'activation': ['relu'],
    'solver': ['adam'],
    'alpha': [0.0001, 0.0005],
    'learning_rate_init': [0.001, 0.005],
    'max_iter': [800]
}


In [14]:
from sklearn.model_selection import GridSearchCV

grid = GridSearchCV(
    MLPClassifier(random_state=42),
    param_grid,
    cv=3,
    scoring='accuracy'
)

grid.fit(X_train, y_train)

print("\n--- GRID SEARCH RESULTS ---")
print("Best Accuracy:", grid.best_score_)
print("Best Parameters:", grid.best_params_)


--- GRID SEARCH RESULTS ---
Best Accuracy: 0.7895021645021645
Best Parameters: {'activation': 'relu', 'alpha': 0.0001, 'hidden_layer_sizes': (64, 32), 'learning_rate_init': 0.001, 'max_iter': 800, 'solver': 'adam'}


### 4. Evaluation
- Employ suitable metrics such as accuracy, precision, recall, and F1-score to evaluate your model's performance.
- Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning.


In [15]:
best_model = grid.best_estimator_
y_pred_tuned = best_model.predict(X_test)

print("\n--- TUNED ANN PERFORMANCE ---")
print("Accuracy :", accuracy_score(y_test, y_pred_tuned))
print("Precision:", precision_score(y_test, y_pred_tuned))
print("Recall   :", recall_score(y_test, y_pred_tuned))
print("F1 Score :", f1_score(y_test, y_pred_tuned))

print("\nClassification Report:")
print(classification_report(y_test, y_pred_tuned))


--- TUNED ANN PERFORMANCE ---
Accuracy : 0.8571428571428571
Precision: 0.9375
Recall   : 0.75
F1 Score : 0.8333333333333334

Classification Report:
              precision    recall  f1-score   support

           0       0.81      0.95      0.88        22
           1       0.94      0.75      0.83        20

    accuracy                           0.86        42
   macro avg       0.87      0.85      0.85        42
weighted avg       0.87      0.86      0.86        42



### Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning.
The default ANN model achieved higher accuracy (90.47%) compared to the tuned model (85.71%). Although hyperparameter tuning explored different network architectures, learning rates, and regularization strengths, the performance slightly decreased across all evaluation metrics. This indicates that the default hyperparameters were already well-suited to the dataset. The tuned model likely introduced slightly stronger regularization or a higher learning rate, which reduced model flexibility and marginally lowered recall and precision. Overall, hyperparameter tuning did not significantly degrade performance but also did not improve it, suggesting that the original configuration was close to optimal for this problem.